kaos/algorithm/visibility_finder.py

```diff
@@ -335,7 +346,54 @@ def determine_visibility(self, error=0.1, tolerance_ratio=0.1, max_iter=100):
         # If the loop terminates and an access end was still not found that means that point should
         # still be visible at the end of the period.
         # NOTE: subinterval_end would also work here but is difficult to test.
-        if access_start is not None:
+        if (access_start is not None) and (access_start < end_time):
+            if self.visibility(end_time) <= 0:
+                raise VisibilityFinderError("Visibility interval started at {} "
+                                            "but did not end at {}".format(access_start, end_time))
+            sat_accesses.append((access_start, end_time))
+
+        # TODO: switch this to log
+        print("average h: {}".format((end_time-start_time)/interval_num))
+
+        return sat_accesses
+
+
+    def determine_visibility_brute_force(self, step=5):
+        """Find visibility intervals using brute-force method. Visibility of site is checked at
+        intervals defined by step.
+
+        Args:
+            step (int): brute-force step, defaults to 5 second as mentioned in the cited paper.
+
+        Returns:
+            The subintervals over which the site is visible.
+        """
+        # Reduce unnecessary accuracy temporarily
+        global_accuracy = mp.mp.dps
+        mp.mp.dps = 5
+
+        real_function = lambda t:(self.visibility(t))
+
+        start_time, end_time = self.interval
+        sat_accesses = []
+        access_start = None
+        for time in range(start_time, end_time, step):
+            Visibility_val = self.visibility(time)
+
+            if access_start == None:
+                if (Visibility_val > 0):
+                    access_start = time
+            else:
+                if (Visibility_val < 0):
+                    sat_accesses.append((access_start, time))
+                    access_start = None
+
+        #Reduce accuracy temporarily
+        mp.mp.dps = global_accuracy
+
+        # If the loop terminates and an access end was still not found that means that point should
+        # still be visible at the end of the period.
+        if (access_start is not None) and (access_start < end_time):
             if self.visibility(end_time) <= 0:
```

Comment:
I think it's correct as it is. It is raising an error when we expect it to have been appended but it is not.

Reply:
visibility(end_time) <= 0 implies that visibility interval has ended. That contradicts the error message!

Reply:
visibility(end_time) <= 0 implies it has ended but the function's return value implies it has not ended. Sure sounds like an error to me!
Also, this has been part of the master since #22; Ask @ZeyadTamimi about it.

Reply:
Ok so let me work though this ( it also applies to the code in the non brute force ):
```if (access_start is not None) and (access_start < end_time):```
Only executes when we have looped through the entire interval and have started a visibility epoch and have not ended it. I cant remember exactly why I put in the extra condition of (access_start < end_time) but that is inconsequential.

Now logically speaking since our algorithm should be acuate we expect that self.visibility(end_time) is still positive since according to the roots we have yet to end the visibility epoch, so if we test it and self.visibility(end_time) shows that we are in a non visible epoch then the application is contradicting itself, hence the error.

Reply:
Thanks! Also just for completeness:
I added the extra condition (access_start < end_time). It is because I was hitting cases where the end of the final subinterval was after the "end_time" (i.e. subinterval_end > end_time) and it was finding a root after the "end_time".
Without the extra check, you would get errors like "Visibility interval started at 150 but did not end at 100" which shouldn't be an error because we don't care about access_start if it is after the end_time; Hence the extra condition.

Reply:
I can see that happening in visibility_finder. However in the brute_force method, would the end of the final subinterval appear after the "end_time", since there is no root finding involved.

Reply:
I'm not sure what you are commenting on...
if it is about ```(access_start is not None)``` check, it can still cause problems in brute-force when end_time is in the middle of a visibility period. i.e. end of visibility is after the end_time.
if it is about ```(access_start < end_time)``` check, as I already said, it doesn't do anything for brute-force (cuz it's always true). It is there to keep the checks the same across the two visibility functions.

Reply:
Sorry I misunderstood. I am commenting about (access_start < end_time). Reading your last post makes me think that you added (access_start < end_time) in brute_force to address the cases where the end of the final subinterval was after the "end_time".

Reply:
My bad... should have mentioned it's intended for the determin_visibility(). It's just copied over to brute-force for symmetry. Let me know if anything else looks weird.